common/containers/Tabs/SendTransaction/index.jsx

```diff
@@ -343,20 +377,20 @@ export class SendTransaction extends React.Component {
   }
 
   async estimateGas() {
-    const trans = await this.getTransactionInfoFromState();
-    if (!trans) {
-      return;
-    }
-
-    // Grab a reference to state. If it has changed by the time the estimateGas
-    // call comes back, we don't want to replace the gasLimit in state.
-    const state = this.state;
-
-    this.props.nodeLib.estimateGas(trans).then(gasLimit => {
+    try {
+      const transaction = await this.getTransactionInfoFromState();
+      // Grab a reference to state. If it has changed by the time the estimateGas
+      // call comes back, we don't want to replace the gasLimit in state.
+      const state = this.state;
+      const gasLimit = await this.props.nodeLib.estimateGas(transaction);
       if (this.state === state) {
         this.setState({ gasLimit: formatGasLimit(gasLimit, state.unit) });
+      } else {
+        this.estimateGas();
```

Comment:
if the state has changed in the meantime, estimateGas again (with the hope that this time the state won't change while the async estimation is being made).

Reply:
Yes, agreed. I'll make an issue to do this before V4 full-launch.

Reply:
I second (third?) this, I could help out with porting a lot of this logic to the sagas.

Reply:
Yes please!